### PR TITLE
Issue #7872: fixes Resolve Pitest Issues - ImportOrderCheck (1)

### PR DESCRIPTION
**Pull request by ImmortalRabbit**
Fixes Issue #7872 

1) Pitest-report Line 903
- Before: https://immortalrabbit.github.io/pit-reports/202004070645/com.puppycrawl.tools.checkstyle.checks.imports/ImportOrderCheck.java.html
- After: https://immortalrabbit.github.io/pit-reports_after/202004070732/com.puppycrawl.tools.checkstyle.checks.imports/ImportOrderCheck.java.html
2) Hardcode: https://github.com/ImmortalRabbit/checkstyle/commit/f78de6e3f05ef3b7559252b418d5cd0177b93e22
3) Diff report: https://immortalrabbit.github.io/reports/diff/

### CODE REVIEW

**Comment by strkkk** [comment]
@ImmortalRabbit you have 2 commits in your branch, please make it 1.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 1]
```diff
@@ -909,7 +909,7 @@ private static int getGroupNumber(Pattern[] patterns, String name) {
                     bestEnd = matcher.end();
                     bestPos = matcher.start();
                 }
-                else if (matcher.start() == bestPos && matcher.end() > bestEnd) {
+                else if (bestEnd < matcher.end()) {
```
Comment (baratali) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:912]: Please make it `matcher.end() > bestEnd`. Looks more readable this way

Comment (ImmortalRabbit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:912]: That's what I used for diff report:

```groovy diff.groovy --localGitRepo ../../../IdeaProjects/checkstyle --baseBranch master --patchBranch issue_7872_hardcode --config my_check.xml --listOfProjects projects-to-test-on.properties```

`cat my_check.xml`
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="TreeWalker">
	<module name="ImportOrder">
  	    <property name="groups" value="*,javax,java"/>
  	    <property name="ordered" value="true"/>
  	    <property name="separated" value="false"/>
  	    <property name="option" value="bottom"/>
  	    <property name="sortStaticImportsAlphabetically" value="true"/>
	</module>
	<module name="SuppressionXpathSingleFilter">
  	    <property name="checks" value="ImportOrder"/>
      	    <property name="message" value="^'java\..*'.*"/>
	</module>
    </module>
    <!-- Example of filter -->
    <!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
    -->

    <!-- as we run on regression even on non-compiled files we need to skip exceptions on them -->
    <module name="SuppressionSingleFilter">
      <property name="message" value="Exception occurred while parsing"/>
      <property name="checks" value="Checker"/>
    </module>
</module>

```

Comment (ImmortalRabbit) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:912]: Do I need to change from:
`<property name="groups" value="*,javax,java"/>`
to
`<property name="groups" value="*,javax, java, java.*"/>` or for something else? 

I don't know any other way of killing mutation without removing this check `matcher.start() == bestPos`. Do you know any other possible solutions to kill mutation?

**Comment by baratali** [comment]
@ImmortalRabbit , hey, sorry for delay. I tested this check and its mutation can be covered without condition removal.
The Check changes its behaviour if multiple groups can be found in an import.
For example:
Properties:
```java
checkConfig.addAttribute("groups", "/java.util/,/java.io/");
```
Input file:
```java
package com.puppycrawl.tools.checkstyle.checks.imports.importorder;

import java.util.java.io.File;
import java.io.File;

public class abc {
}
```

Original - no warnings `Audit done.`
After your fix - new warning `5:1: Wrong order for 'java.io.File' import.`

So, you should not delete that condition. You need to write a test using above input file and mutation coverage will increase (when pitest tries to delete that condition the test will fail).

**Comment by baratali** [comment]
@ImmortalRabbit , are you going to continue this fix? Please let us know.

**Comment by ImmortalRabbit** [comment]
> @ImmortalRabbit , are you going to continue this fix? Please let us know.

Sorry, I think I cannot finish it. I am closing my pull requests.

**Pull request closed without merging**